### PR TITLE
[SOFT-188] SPV1020 MPPT driver

### PR DESCRIPTION
**Pull request by copperium**
Supports issuing each command to the SPV1020s and extracting data from the status byte.

Did you know you can have both x86/stm32-specific C files and a common one?

### CODE REVIEW

File: libraries/ms-drivers/src/spv1020_mppt.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include "spv1020_mppt.h"
+#include "spv1020_mppt_defs.h"
+
+// This file implements common functionality between stm32 and x86. See stm32f0xx/spv1020_mppt.c
```
Comment (AveryChiu64) [inline libraries/ms-drivers/src/spv1020_mppt.c:4]: Hmm, why is this file needed if it is neither for x86 or stm32?

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+// Send a command to SPV1020 and get a response. |rx_len| must be <= 2.
+// It's safe to pass NULL as |rx_data| if |rx_len| is 0.
+static StatusCode prv_send_command(SpiPort port, uint8_t command, uint8_t *rx_data, size_t rx_len) {
+  if (port >= NUM_SPI_PORTS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // We don't use spi_exchange since that assumes 0x00 is NOP, whereas SPV1020 appears to use 0x01.
+  spi_cs_set_state(port, GPIO_STATE_LOW);
```
Comment (AveryChiu64) [inline libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:14]: Why is the GPIO_STATE_LOW set twice for the same port?

File: libraries/ms-drivers/test/test_spv1020_mppt.c [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  spi_init(TEST_SPI_PORT, &spi_settings);
+}
+void teardown_test() {}
+
+// Test that we can successfully send the SHUT and Turn ON commands.
+void test_shut_and_turn_on(void) {
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_turn_on(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));  // 2 in a row
```
Comment (AveryChiu64) [inline libraries/ms-drivers/test/test_spv1020_mppt.c:38]: Wait if it is already shut why are they allowed to keep sending this command?

File: libraries/ms-drivers/inc/spv1020_mppt.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+// be set to a 4-bit bitmask indicating which, if any, branches have overcurrent.
+bool spv1020_is_overcurrent(uint8_t status, uint8_t *ovc_branches);
+
+// Return whether the overvoltage (OVV) bit is set in the status byte.
+bool spv1020_is_overvoltage(uint8_t status);
+
+// Return whether the overtemperature (OVT) bit is set in the status byte.
+bool spv1020_is_overtemperature(uint8_t status);
+
+// Return whether the CR bit is set in the status byte. We don't know what the CR bit is, but it
```
Comment (AveryChiu64) [inline libraries/ms-drivers/inc/spv1020_mppt.h:50]: Wait what does CR stand for?

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c (hunk 2 above)
Comment (copperium) [reply libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:14]: Whoops, fixed that

File: libraries/ms-drivers/src/spv1020_mppt.c (hunk 1 above)
Comment (copperium) [reply libraries/ms-drivers/src/spv1020_mppt.c:4]: That file has common functionality, the `spv1020_is_X` functions are the same between x86 and stm32.

File: libraries/ms-drivers/inc/spv1020_mppt.h (hunk 4 above)
Comment (copperium) [reply libraries/ms-drivers/inc/spv1020_mppt.h:50]: It shows up in the datasheet, I don't know what it stands for and neither does Micah, but it might be good to log it.

File: libraries/ms-drivers/test/test_spv1020_mppt.c (hunk 3 above)
Comment (copperium) [reply libraries/ms-drivers/test/test_spv1020_mppt.c:38]: I suppose they aren't supposed to, but since the module doesn't keep state I guess this test is to make sure nothing funky happens if it's called multiple times. I'll add a warning when shut is called when it's already shut on x86.

File: libraries/ms-drivers/src/spv1020_mppt.c (hunk 1 above)
Comment (AveryChiu64) [reply libraries/ms-drivers/src/spv1020_mppt.c:4]: Oh I see, so that saves you the from having to make the two files with duplicate functions

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// This file implements stm32-specific functionality. See ../spv1020_mppt.c for the rest.
+
+// Send a command to SPV1020 and get a response. |rx_len| must be <= 2.
+// It's safe to pass NULL as |rx_data| if |rx_len| is 0.
+static StatusCode prv_send_command(SpiPort port, uint8_t command, uint8_t *rx_data, size_t rx_len) {
+  if (port >= NUM_SPI_PORTS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // We don't use spi_exchange since that assumes 0x00 is NOP, whereas SPV1020 appears to use 0x01.
```
Comment (g-jessmuir) [inline libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:13]: I'm not sure this is correct. In the command list NOP is 0x01, but on page 12 paragraph 3 it says:
> Therefore the master appends a number of NOPs (0x00) to the command so that the entire data can be transmitted

It makes a lot more sense to me for this to be left as 0x00, matching spi_exchange. I think we should request a `smoke_spi` test with exchange before jumping into this weird custom spi implementation.

My theory is that the *command word* NOP is different from the SPI NOP, and that the SPI NOP is really 0x00.

I hate this datasheet.

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+StatusCode spv1020_read_current(SpiPort port, uint16_t *current) {
+  return prv_send_command(port, COMMAND_READ_CURRENT, (uint8_t *)current, 2);
+}
+
+StatusCode spv1020_read_voltage_in(SpiPort port, uint16_t *vin) {
+  return prv_send_command(port, COMMAND_READ_VIN, (uint8_t *)vin, 2);
+}
+
+StatusCode spv1020_read_pwm(SpiPort port, uint16_t *pwm) {
+  return prv_send_command(port, COMMAND_READ_PWM, (uint8_t *)pwm, 2);
```
Comment (g-jessmuir) [inline libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:43]: I think it makes sense to convert the PWM to the actual value here rather than the steps since it requires no calibration, but if you'd rather leave that to the user I'm ok with it. In other places we use uints as decimal by using 123.4% = `1234`.

File: libraries/ms-drivers/test/test_spv1020_mppt.c [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  TEST_ASSERT_BITS_LOW(0b1111110000000000, vin);
+}
+
+// Test that we can successfully read PWM and it's in the stated range.
+void test_read_pwm(void) {
+  uint16_t pwm = 0xFFFF;  // over 9 bits
+  TEST_ASSERT_OK(spv1020_read_pwm(TEST_SPI_PORT, &pwm));
+
+  // make sure the 6 highest bits are low, so it's a 9-bit value
+  TEST_ASSERT_BITS_LOW(0b1111111000000000, pwm);
```
Comment (g-jessmuir) [inline libraries/ms-drivers/test/test_spv1020_mppt.c:67]: could add an assertion that pwm is less than expected number of steps rather than just a 9 bit number, depends on if you add the conversion to actual pwm from the 9 bits.

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  return prv_send_command(port, COMMAND_READ_CURRENT, (uint8_t *)current, 2);
+}
+
+StatusCode spv1020_read_voltage_in(SpiPort port, uint16_t *vin) {
+  return prv_send_command(port, COMMAND_READ_VIN, (uint8_t *)vin, 2);
+}
+
+StatusCode spv1020_read_pwm(SpiPort port, uint16_t *pwm) {
+  uint16_t raw_pwm;
+  StatusCode code = prv_send_command(port, COMMAND_READ_PWM, (uint8_t *)raw_pwm, 2);
```
Comment (g-jessmuir) [inline libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:44]: I believe this should be (uint8_t *)&raw_pwm, might need more brackets

File: libraries/ms-drivers/inc/spv1020_mppt_defs.h [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+#define COMMAND_READ_CURRENT 0x04
+#define COMMAND_READ_VIN 0x05
+#define COMMAND_READ_PWM 0x06
+#define COMMAND_READ_STATUS 0x07
+
+// Bitmasks for extracting information from the status byte.
+#define OVC_BITMASK 0b1111000
+#define OVV_BITMASK 0b0000100
+#define OVT_BITMASK 0b0000010
+#define CR_BITMASK 0b0000001
```
Comment (GeraldNDA) [inline libraries/ms-drivers/inc/spv1020_mppt_defs.h:18]: Typically in driver code (and in general) we prefix these. In thise case I guess it'd be `SPV1020_` may or may not want to shorten BITMASK to MASK and COMMAND to CMD to compensate.

File: libraries/ms-drivers/src/x86/spv1020_mppt.c [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+static bool s_is_shut = false;
+
+StatusCode spv1020_shut(SpiPort port) {
+  if (port >= NUM_SPI_PORTS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  // this doesn't occur on stm32 but is useful for debugging
+  if (s_is_shut) {
+    LOG_WARN("Issuing SHUT command to SPV1020 MPPT when it's already shut\r\n");
```
Comment (GeraldNDA) [inline libraries/ms-drivers/src/x86/spv1020_mppt.c:22]: Why carriage returns?

Comment (copperium) [reply libraries/ms-drivers/src/x86/spv1020_mppt.c:22]: Minicom needs carriage returns to move to the beginning of the new line like a typewriter. There's a setting to auto-insert carriage returns, but that was the rational behind putting them in. (I suppose this is x86 so we aren't using minicom anyways, I'll remove them.)

File: libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c (hunk 8 above)
Comment (copperium) [reply libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c:44]: Whoops, fixed
